src/transxchange/read.rs

```diff
@@ -378,90 +378,241 @@ impl Iterator for NaiveDateRange {
+        // but, New Year and New Year Bank Holiday are 2 different things:
+        // - New Year is always on 1st of January
+        // - New Year Bank Holiday might be on the 2nd or 3rd if the 1st ends
+        // up being on a weekend
+        //
+        // TransXChange has a notion for both New Year and New Year Bank Holiday
+        // so we also need to add the New Year days in `bank_holidays`.
+        // The above explanation applies to:
+        // - New Year (1st of January)
+        // - 2nd of January (specific to Scotland)
+        // - Saint-Andrews (30th of November, specific to Scotland)
+        // - Christmas (25th of December)
+        // - Boxing Day (26th of December)
+        let mut bank_holidays = bank_holidays.clone();
+        let new_year_days = bank_holidays::get_fixed_days(1, 1, validity_period);
+        bank_holidays.insert(BankHoliday::NewYear, new_year_days);
+        let january_second_days = bank_holidays::get_fixed_days(2, 1, validity_period);
+        bank_holidays.insert(BankHoliday::JanuarySecond, january_second_days);
+        let saint_andrews_days = bank_holidays::get_fixed_days(30, 11, validity_period);
+        bank_holidays.insert(BankHoliday::SaintAndrews, saint_andrews_days);
+        let christmas_days = bank_holidays::get_fixed_days(25, 12, validity_period);
+        bank_holidays.insert(BankHoliday::Christmas, christmas_days);
+        let boxing_days = bank_holidays::get_fixed_days(26, 12, validity_period);
+        bank_holidays.insert(BankHoliday::BoxingDay, boxing_days);
+        let days_included =
+            if let Ok(operation) = bank_holiday_operation.try_only_child("DaysOfOperation") {
+                get_bank_holidays_dates(operation, &bank_holidays)
+            } else {
+                Default::default()
+            };
+        let days_excluded =
+            if let Ok(operation) = bank_holiday_operation.try_only_child("DaysOfNonOperation") {
+                get_bank_holidays_dates(operation, &bank_holidays)
+            } else {
+                Default::default()
+            };
+        Ok((days_included, days_excluded))
     }
+    let mut dates = BTreeSet::new();
+    let regular_days = get_regular_days(operating_profile.try_only_child("RegularDayType")?)?;
+    let bank_holidays = operating_profile
+        .try_only_child("BankHolidayOperation")
+        .and_then(|bank_holiday_operation| {
+            get_bank_holidays(bank_holiday_operation, bank_holidays, validity_period)
+        })
+        .ok();
     for date in NaiveDateRange::new(validity_period.start_date, validity_period.end_date) {
-        if days.contains(&date.weekday()) {
-            // TODO: Handle exceptions (see SpecialDaysOperation)
-            // TODO: Handle bank holidays (see BankHolidaysOperation)
+        let is_regular_day = regular_days.contains(&date.weekday());
+        let is_bank_holiday = if let Some((ref included_bank_holidays, _)) = bank_holidays {
```

Comment:
```rust
let is_bank_holiday = bank_holidays
            .as_ref()
            .map(|(ref included_bank_holidays, _)| included_bank_holidays.contains(&date))
            .unwrap_or_else(|_| false);
        let is_excluded_bank_holiday = bank_holidays
            .as_ref()
            .map(|(_, ref excluded_bank_holidays)| excluded_bank_holidays.contains(&date))
            .unwrap_or_else(|_| false);
```

Reply:
I really refactored that part so it doesn't really apply anymore but thanks for the alternative. These are the case where I'm not sure which one is the most readable though.